File: otp_protection/lib/core/services/protection _engine.dart
import 'package:otp_protection/core/models/protection_event.dart';
import 'package:otp_protection/core/models/sms_message.dart';
import 'package:otp_protection/core/services/call_detector.dart';
import 'package:otp_protection/core/services/event_puplisher.dart';
import 'package:otp_protection/core/services/message_parser.dart';

class ProtectionEngine {
  final MessageParser parser;
  final CallDetector callDetector;
  final EventPublisher publisher;

  ProtectionEngine({
    required this.parser,
    required this.callDetector,
    required this.publisher,
  });

  void handleSms(SmsMessage sms) {
    final result = parser.parse(sms);

    if (!result.containsOtp) return;

    final event = callDetector.isUserInCall
        ? ProtectionEvent.otpReceivedDuringCall
        : ProtectionEvent.otpDetected;

    publisher.publish(event);
  }
}

File: otp_protection/android/app/src/main/kotlin/com/example/otp_protection/protection_engine.kt
package com.example.otp_protection
import android.content.Context

class ProtectionEngine(
    private val parser: SmsParser,
    private val callDetector: CallDetector,
    private val publisher: EventPublisher
) {

    fun handleSms(
        context: Context,
        message: String
    ) {

        if (!parser.containsOtp(message)) {
            return
        }

        val event =
            if (
                callDetector.isUserInCall(
                    context
                )
            ) {
                ProtectionEvent.otpReceivedDuringCall
            } else {
                ProtectionEvent.otpDetected
            }

        publisher.publish(event)
    }
}
